File: mobile/android/app/src/main/kotlin/dev/activechain/wallet/OpenId4VpDeepLink.kt
package dev.activechain.wallet

import java.net.URI

data class OpenId4VpReviewIntent(
    val requestUri: URI,
    val clientId: String,
    val nonce: String,
    val state: String,
)

/** Deep links can only create a wallet-owned review intent; they cannot approve or post. */
object OpenId4VpDeepLink {
    fun parse(raw: String): OpenId4VpReviewIntent {
        val uri = URI(raw)
        require(uri.scheme == "openid4vp" && uri.host == "authorize" && uri.fragment == null)
        val query = mutableMapOf<String, String>()
        uri.rawQuery.orEmpty().split('&').filter { it.isNotEmpty() }.forEach { part ->
            val pair = part.split('=', limit = 2)
            require(pair.size == 2 && pair[0] !in setOf("redirect_uri", "response_uri"))
            val key = java.net.URLDecoder.decode(pair[0], Charsets.UTF_8)
            require(key !in query)
            query[key] = java.net.URLDecoder.decode(pair[1], Charsets.UTF_8)
        }
        val request = URI(query.getValue("request_uri"))
        require(request.scheme == "https" && request.userInfo == null && request.fragment == null)
        return OpenId4VpReviewIntent(request, query.getValue("client_id"), query.getValue("nonce"), query.getValue("state"))
    }
}

File: mobile/apple/AmberApp/Sources/OpenId4VpDeepLink.swift
import Foundation

struct OpenId4VpReviewIntent: Equatable {
    let requestURI: URL
    let clientID: String
    let nonce: String
    let state: String
}

enum OpenId4VpDeepLink {
    static func parse(_ url: URL) throws -> OpenId4VpReviewIntent {
        guard url.scheme == "openid4vp", url.host == "authorize", url.fragment == nil,
              let parts = URLComponents(url: url, resolvingAgainstBaseURL: false) else { throw ParseError.invalid }
        var query: [String: String] = [:]
        for item in parts.queryItems ?? [] {
            guard let value = item.value, query[item.name] == nil,
                  item.name != "redirect_uri", item.name != "response_uri" else { throw ParseError.invalid }
            query[item.name] = value
        }
        guard let raw = query["request_uri"], let request = URL(string: raw),
              request.scheme == "https", request.user == nil, request.fragment == nil,
              let client = query["client_id"], let nonce = query["nonce"], let state = query["state"]
        else { throw ParseError.invalid }
        return OpenId4VpReviewIntent(requestURI: request, clientID: client, nonce: nonce, state: state)
    }
    enum ParseError: Error { case invalid }
}
